cmd/server/pactasrv/analysis.go

```diff
@@ -341,3 +283,88 @@ func (s *Server) RunAnalysis(ctx context.Context, request api.RunAnalysisRequest
 
 	return api.RunAnalysis200JSONResponse{AnalysisId: string(result)}, nil
 }
+
+func (s *Server) analysisDoAuthzAndAuditLog(ctx context.Context, analysisID pacta.AnalysisID, action pacta.AuditLogAction) error {
+	actorInfo, err := s.getActorInfoOrFail(ctx)
+	if err != nil {
+		return err
+	}
+	analysis, err := s.DB.Analysis(s.DB.NoTxn(ctx), analysisID)
+	if err != nil {
+		if db.IsNotFound(err) {
+			return notFoundOrUnauthorized(actorInfo, action, pacta.AuditLogTargetType_Analysis, analysisID)
+		}
+		return oapierr.Internal("querying analysis for authz failed", zap.Error(err))
+	}
+	as := &authzStatus{
+		primaryTargetID:      string(analysisID),
+		primaryTargetType:    pacta.AuditLogTargetType_Analysis,
+		primaryTargetOwnerID: analysis.Owner.ID,
+		actorInfo:            actorInfo,
+		action:               action,
+	}
+	switch action {
+	case pacta.AuditLogAction_Update, pacta.AuditLogAction_Delete, pacta.AuditLogAction_ReadMetadata:
+		as.isAuthorized, as.authorizedAsActorType = allowIfAdminOrOwner(actorInfo, analysis.Owner.ID)
```

Comment:
I think the approach of handling authz and audit logging at the same time is fine, but using the audit logging _verbs_ for authz seems like muddying the concepts too far. I'm not suggesting changing it, but I'd just be aware of it as you go forward, I can imagine areas where this could be problematic.

If you hit those cases, I'd just separate out the authz verb from the audit log verb.

Reply:
I went back on this SO MANY TIMES lol. I tend to agree with you, particularly for big systems, couplings between auth and logging are bound to cause issues. However, the thing I kept coming back to was accidental drift: thinking "hm what audit verb should this be" for an authz verb that didn't fit precicely into an existing one. I eventually came to the conclusion that if we have two sets, we invite drift between the idea of "check authorization" and "log successful authorization", and after a few such accidents, I realized this is probably less error prone for a (relatively) small project like this one.